### DISCUSSION TITLE
Supporting ESP32_S3 using W5500 Ethernet

### DISCUSSION THREAD

**Discussion by avanderhoest** [khoih-prog/EthernetWebServer#72] [Q&A]
Hey! Thanks for the lib!
I used the example [ESP32_FS_EthernetWebserver](https://github.com/khoih-prog/EthernetWebServer/tree/master/examples/ESP32_FS_EthernetWebServer) and it works when I run it from Arduino IDE.
When I upload the code from PlatformIO, I get the correct logging :
   Using mac index = 17
   Connected! IP address: 192.168.1.221
   FS File: favicon.ico, size: 1.12KB
   ...
   HTTP server started @192.168.1.221
   Open http://192.168.1.221/edit to see the file browser
   handleFileRead: /

But my browser doesn't show the html pages. It is stuck on loading.

I use the same example code. Do you have any idea why code uploaded by PlatformIO doesn't serve the files?

ps. My setup is an WT32-SC01-PLUS board and I use an ethernet W5500 module. 
Should I be using your [WebServer_WT32_ETH01](https://github.com/khoih-prog/WebServer_WT32_ETH01) lib instead?

**Comment by khoih-prog** [comment]
Hi @avanderhoest 

> I used the example ESP32_FS_EthernetWebserver and it works when I run it from Arduino IDE.

and

> code uploaded by PlatformIO doesn't serve the files

clearly show that there is some kind of issue with your settings, even some bug of PIO.

I suggest you post on PIO Forum to ask for help.

Good Luck,

**Reply by avanderhoest** [reply]
Wow thanks for your quick reply. I was even still updating my original question with the question if I should be using [WebServer_WT32_ETH01](https://github.com/khoih-prog/WebServer_WT32_ETH01) instead of [ESP32_FS_EthernetWebserver](https://github.com/khoih-prog/EthernetWebServer/tree/master/examples/ESP32_FS_EthernetWebServer)?


Also, do you have a lib or example to use (AsyncWebsever)[https://github.com/me-no-dev/ESPAsyncWebServer] together with an  ESP32 with Ethernet?

**Reply by khoih-prog** [reply]
For ESP32 Ethernet Async, use

1. [**AsyncWebServer_WT32_ETH01**](https://github.com/khoih-prog/AsyncWebServer_WT32_ETH01) for **WT32_ETH01** (ESP32 + LAN8720)
2. [**AsyncWebServer_ESP32_ENC**](https://github.com/khoih-prog/AsyncWebServer_ESP32_ENC) for ESP32_ENC (**ESP32 + ENC28J60**)

It's better to use Async, especially for ESP32 and ESP8266, if you have experience and expect higher performance. Check [**Why Async is better**](https://github.com/khoih-prog/AsyncWebServer_WT32_ETH01#why-async-is-better)

**Reply by avanderhoest** [reply]
Thank you! 
Do you also have a suggestion for WT32 with Ethernet W5500? 
(I have WT32-SC01-PLUS)

**Comment by khoih-prog** [comment]
Try my `brand new`

1. [**WebServer_ESP32_W5500**](https://github.com/khoih-prog/WebServer_ESP32_W5500) for sync
2. [**AsyncWebServer_ESP32_W5500**](https://github.com/khoih-prog/AsyncWebServer_ESP32_W5500) for Async

libraries written for ESP32 + W5500

**Reply by avanderhoest** [reply]
amazing! I will try them!

**Reply by avanderhoest** [reply]
Hi Khoih!

I tried the Async webserver but no luck so far. I get:

```
  Start AsyncSimpleServer_ESP32_W5500 on TINYS3 with ESP32_W5500
 AsyncWebServer_ESP32_W5500 v1.6.2 for core v2.0.0+
 [AWS] Default SPI pinout:
 [AWS] SPI_HOST: 2
 [AWS] MOSI: 14
 [AWS] MISO: 13
 [AWS] SCK: 12
 [AWS] CS: 11
 [AWS] INT: 10
 [AWS] SPI Clock (MHz): 25
 [AWS] =========================
 [AWS] esp_eth_mac_new_esp32 failed
```

I am quite sure about most PIN numbers because I got your ESP32_FS_EthernetWebServer example working:

```
Starting ESP32_FS_EthernetWebServer on ESP32 using SPIFFS
With W5x00 using Ethernet_Generic Library on SPI
EthernetWebServer v2.2.3
[EWS] =========== USE_ETHERNET_GENERIC ===========
[EWS] Default SPI pinout:
[EWS] MOSI: 14
[EWS] MISO: 13
[EWS] SCK: 12
[EWS] SS: 22
[EWS] =========================
[EWS] ESP32 setCsPin: 11
=========================
Currently Used SPI pinout
MOSI:14
MISO:13
SCK:12
CS/SS:11
=========================
Using mac index = 15
Connected! IP address: 192.168.1.221
FS File: favicon.ico, size: 1.12KB
FS File: index.htm, size: 3.63KB
FS File: graphs.js.gz, size: 1.92KB
FS File: edit.htm.gz, size: 4.02KB
FS File: CanadaFlag_1.png, size: 40.25KB
FS File: CanadaFlag_2.png, size: 8.12KB
FS File: CanadaFlag_3.jpg, size: 10.89KB

HTTP server started @192.168.1.221
```

I connected INT to the display pin next to CS. But I don't know about the SPI_HOST and the SPI Clock.

Also, I would need a local network with fixed IP so I think I need:

```
IPAddress myIP(192, 168, 1, 221);
IPAddress myGW(192, 168, 1, 1);
IPAddress mySN(255, 255, 255, 0);

// Google DNS Server IP
IPAddress myDNS(8, 8, 8, 8);
```

and :

```
  // start the ethernet connection and the server:
  // Use DHCP dynamic IP and random mac
  uint16_t index = millis() % NUMBER_OF_MAC;

  //bool begin(int MISO_GPIO, int MOSI_GPIO, int SCLK_GPIO, int CS_GPIO, int INT_GPIO, int SPI_CLOCK_MHZ,
  //           int SPI_HOST, uint8_t *W5500_Mac = W5500_Default_Mac);
  //ETH.begin( MISO_GPIO, MOSI_GPIO, SCK_GPIO, CS_GPIO, INT_GPIO, SPI_CLOCK_MHZ, ETH_SPI_HOST );
  ETH.begin( MISO_GPIO, MOSI_GPIO, SCK_GPIO, CS_GPIO, INT_GPIO, SPI_CLOCK_MHZ, ETH_SPI_HOST, mac[index] );

  // Static IP, leave without this line to get IP via DHCP
  bool config(IPAddress local_ip, IPAddress gateway, IPAddress subnet, IPAddress dns1 = 0, IPAddress dns2 = 0);
  ETH.config(myIP, myGW, mySN, myDNS);

  ESP32_W5500_waitForConnect();
```

but then I get an error saying:

```
conversion from 'int' to 'IPAddress' is ambiguous
   bool config(IPAddress local_ip, IPAddress gateway, IPAddress subnet, IPAddress dns1 = 0, IPAddress dns2 = 0);
```
                                                                                                              
Any pointers to how to continue? Thanks in advance!!

**Comment by khoih-prog** [comment]
HI @avanderhoest 

I have no compile error issue with static IP turned on for **ESP32_S3** or even **UM_TinyS3**. So be sure to use the latest ESP32 core **v2.0.5** and set compiler warning to `Default` (For Arduino IDE, **File->Preferences->Compiler Warnings**) as recent ESP32 core / Arduino IDE is so extreme to flag even common warnings as errors

![Selection_202](https://user-images.githubusercontent.com/57012152/205676960-501faa8f-5cce-403d-a194-5b250edf8cd5.png)

![Selection_203](https://user-images.githubusercontent.com/57012152/205677310-0b17cbbc-728c-427c-9d18-7d605946e49f.png)

2. The INT pin is absolutely necessary to use the library.
So be sure to use it correctly. You can change **INT** from pin 10 (commonly used by many other libraries / parts) to another in to test and be sure it not conflicting.


![image](https://user-images.githubusercontent.com/57012152/205680745-41880794-fba1-4d97-8706-1e60476a0e02.png)


![Selection_204](https://user-images.githubusercontent.com/57012152/205681582-1521dc5f-ed32-4b9c-9a4d-23502855e433.png)


#### Note:

**Don't use GPIO10, change to another GPIO, such as IO4, IO5, etc.**

Also it's better to try using **Dynamic IP** first, if you have issue with **Static IP**.

Only you can know and check the wiring to be sure correct.

**Comment by avanderhoest** [comment]
Hey! I am trying with Dynamic IP first indeed, but still no luck.

I changed INT from GPIO10 to GPIO15 but same result.

```
 Start AsyncSimpleServer_ESP32_W5500 on TINYS3 with ESP32_W5500
 AsyncWebServer_ESP32_W5500 v1.6.2 for core v2.0.0+
 Default SPI pinout:
 SPI_HOST: 3
 MOSI: 14
 MISO: 13
 SCK: 12
 CS: 11
 INT: 15
 SPI Clock (MHz): 25
 =========================
 esp_eth_mac_new_esp32 failed
```

I can't use other pins because I have the display, so my only free ones are 10 or 15. This is my setup:
![IMG_2376 copy](https://user-images.githubusercontent.com/25819799/205970149-f9f5f8af-de0a-4d70-ae1e-f59aea403ff4.jpg)

The thing is with these pin numbers I got your example of ESP32_FS_EthernetWebServer working.

Is there any way I can debug more? Like get some more information on "esp_eth_mac_new_esp32 failed".

**Comment by khoih-prog** [comment]
1. Try with other different ESP32 boards to see the W5500 is working. The INT is absolutely necessary. The `ESP32_FS_EthernetWebServer` doesn't need INT to work
2. Buy new W5500, many different styles, to test and see which has working INT
3. Be sure the GPIO15 is not used be anything else

Sorry nobody can help you out here. Try to be investigative and innovative.

**Reply by avanderhoest** [reply]
Thank you! 

I tried another esp32 board and got it working! so I'll try and figure out what was wrong with my original board. Thanks so much!

**Reply by avanderhoest** [reply]
Is the lib also suited for esp32 S3 boards? I would still very much like to make it work on my WT32-SC01-PLUS but so far no luck. Can I leave the Reset pin empty?

When I connect my W5500 to the display board, I can get your [ESP32_FS_EthernetWebserver](https://github.com/khoih-prog/EthernetWebServer/tree/master/examples/ESP32_FS_EthernetWebServer) example working.

When I connect also the interrupt pin to run the [AsyncWebServer_ESP32_W5500](https://github.com/khoih-prog/AsyncWebServer_ESP32_W5500) code, then I get an error "esp_eth_mac_new_esp32 failed". I tried using PIN number 13 or 10. Both of these pins work when I use them for MISO in the  [ESP32_FS_EthernetWebserver](https://github.com/khoih-prog/EthernetWebServer/tree/master/examples/ESP32_FS_EthernetWebServer)  code, so I assume they should be OK. 

What is your setup with connections to the W5500 when you run the  [AsyncWebServer_ESP32_W5500](https://github.com/khoih-prog/AsyncWebServer_ESP32_W5500) code? I leave RST, 1 GND and 5V open. I use the other GND and 3.3V.

**Comment by khoih-prog** [comment]
Hi @avanderhoest 

Unfortunately, this [**EthernetWebServer**](https://github.com/khoih-prog/EthernetWebServer) library as well as many more, such as

- [**WebServer_ESP32_W5500**](https://github.com/khoih-prog/WebServer_ESP32_W5500)
- [**WebServer_ESP32_ENC**](https://github.com/khoih-prog/WebServer_ESP32_ENC)
- even in ESP32 core

don't support new **ESP32_S3, ESP32_S2 and ESP32_C3** using **LwIP** Ethernet

**OK to use with non-LwIP [**Ethernet_Generic**](https://github.com/khoih-prog/Ethernet_Generic) library**

I had to do some more researches, tests and figured out a fix and will create the new library to support specifically **ESP32_S3, , ESP32_S2 and ESP32_C3** and will published soon

---

### Debug terminal

```
Start AdvancedWebServer on ESP32S3_DEV with ESP32_S3_W5500
WebServer_ESP32_S3_W5500 v1.5.1 for core v2.0.0+
[EWS] Default SPI pinout:
[EWS] SPI_HOST: 2
[EWS] MOSI: 11
[EWS] MISO: 13
[EWS] SCK: 12
[EWS] CS: 10
[EWS] INT: 4
[EWS] SPI Clock (MHz): 25
[EWS] =========================

ETH Started
ETH Connected
ETH MAC: DE:AD:BE:EF:BE:0E, IPv4: 192.168.2.106
FULL_DUPLEX, 100Mbps
HTTP EthernetWebServer is @ IP : 192.168.2.106
.......... .......... .......... .
```



![Selection_217](https://user-images.githubusercontent.com/57012152/207437290-5e13d229-9463-4e5f-9db1-360dfbb85f5d.png)


![Selection_215](https://user-images.githubusercontent.com/57012152/207437843-5e910552-5c38-401f-8a71-deb5f6f9ed0c.png)

**Comment by khoih-prog** [comment]
Hi @avanderhoest 


The new [**WebServer_ESP32_SC_W5500 releases v1.0.0**](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/releases/tag/v1.0.0) has just been published. Your contribution is noted in [**Contributions and Thanks**](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500#contributions-and-thanks)

Please test and report [**issue**](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/issues) in that library

Best Regards,

---

#### Releases v1.0.0

1. Initial coding to support ESP32_S3-based boards using `LwIP W5500 Ethernet`.
2. Use `allman astyle`

**Comment by khoih-prog** [comment]
Hi @avanderhoest 

FYI, ESP32_S3 is working now with ENC28J60

The new [**WebServer_ESP32_SC_ENC releases v1.0.0**](https://github.com/khoih-prog/WebServer_ESP32_SC_ENC/releases/tag/v1.0.0) has just been published. Your contribution is also noted in [**Contributions and Thanks**](https://github.com/khoih-prog/WebServer_ESP32_SC_ENC#contributions-and-thanks)

Please test and report [**issue**](https://github.com/khoih-prog/WebServer_ESP32_SC_ENC/issues) in that library

Best Regards,

---

#### Releases v1.0.0

1. Initial coding to support ESP32_S3-based boards using `LwIP ENC28J60 Ethernet`.
2. Use `allman astyle`

**Reply by avanderhoest** [reply]
Hi Khoih! Thanks so much for your quick replies, and all your work!
My ESP32_S3 already worked with your webserver library [EthernetWebServer](https://github.com/khoih-prog/EthernetWebServer), but the issue is with the Async library [AsyncWebServer_ESP32_W5500](https://github.com/khoih-prog/AsyncWebServer_ESP32_W5500/).

Is it possible to fix the issue there also?

**Reply by khoih-prog** [reply]
You can't use current [**AsyncWebServer_ESP32_W5500**](https://github.com/khoih-prog/AsyncWebServer_ESP32_W5500) for ESP32_S3, which strangely behaves differently from ESP32 and can't use the current LwIP library, such as [**WebServer_ESP32_W5500**](https://github.com/khoih-prog/WebServer_ESP32_W5500).

If necessary, I have to modify or write some new libraries, such as AsyncWebServer_ESP32_SC_W5500, specifically for ESP32_S3/2 and C3

Try this new [**AsyncESP32_SC_Ethernet_Manager**](https://github.com/khoih-prog/AsyncESP32_SC_Ethernet_Manager) to see if working for your ESP32_S3 before I'm moving forward.

**Reply by avanderhoest** [reply]
I tried and this is my output:

```
17:31:49.374 -> 
17:31:49.374 -> Starting Async_ESP32_FSWebServer using LittleFS on TINYS3 with ESP32_S3_W5500
17:31:49.374 -> AsyncESP32_SC_Ethernet_Manager v1.0.0
17:31:54.030 -> 
17:31:54.030 -> [EM] Default SPI pinout:
17:31:54.030 -> [EM] SPI_HOST: 2
17:31:54.030 -> [EM] MOSI: 14
17:31:54.030 -> [EM] MISO: 21
17:31:54.030 -> [EM] SCK: 12
17:31:54.030 -> [EM] CS: 11
17:31:54.030 -> [EM] INT: 13
17:31:54.030 -> [EM] SPI Clock (MHz): 25
17:31:54.030 -> [EM] =========================
17:31:54.030 -> 
17:31:54.030 -> ETH Started
17:31:54.104 -> [EM] RFC925 Hostname = AsyncESP32-FSWebServer
17:31:54.104 -> [EM] setSTAStaticIPConfig
17:31:54.104 -> [EM] Set CORS Header to :  Your Access-Control-Allow-Origin
17:31:54.104 -> [EM] LoadCfgFile 
17:31:54.104 -> [EM] failed
17:31:54.104 -> [EM] stationIP = 0.0.0.0 , gatewayIP = 0.0.0.0
17:31:54.104 -> [EM] netMask = 0.0.0.0
17:31:54.104 -> [EM] dns1IP = 0.0.0.0 , dns2IP = 0.0.0.0
17:31:56.052 -> ETH Connected

```

I think it's not enough output, it should also say local ip etc .

These PIN number work with the [EthernetWebServer](https://github.com/khoih-prog/EthernetWebServer) example so they should be correct.

**Comment by khoih-prog** [comment]
Hi @avanderhoest 

I'm moving this to the correct location @ [**Supporting TINYS3 using W5500 Ethernet** #4](https://github.com/khoih-prog/AsyncESP32_SC_Ethernet_Manager/discussions/4)

---

Try to use pin assignment defined in the board package [**um_tinys3** pins_arduino.h](https://github.com/espressif/arduino-esp32/blob/master/variants/um_tinys3/pins_arduino.h)

![Selection_221](https://user-images.githubusercontent.com/57012152/207929078-ea13423f-2cd7-4614-8fcf-57226baf4e7b.png)

```
static const uint8_t SS    = 34;
static const uint8_t MOSI  = 35;
static const uint8_t MISO  = 37;
static const uint8_t SCK   = 36;
```

by changing these lines

https://github.com/khoih-prog/AsyncESP32_SC_Ethernet_Manager/blob/cf92d4baea472bc54da7e2cb28c6a0ad7624159a/examples/Async_ESP32_FSWebServer/Async_ESP32_FSWebServer.ino#L84-L96

```
  // For W5500
  // Optional values to override default settings
  // Don't change unless you know what you're doing
  //#define ETH_SPI_HOST        SPI3_HOST
  //#define SPI_CLOCK_MHZ       25
  
  // Must connect INT to GPIOxx or not working
  //#define INT_GPIO            4
  
  //#define MISO_GPIO           13
  //#define MOSI_GPIO           11
  //#define SCK_GPIO            12
  //#define CS_GPIO             10
```

to


```
  // For W5500
  // Optional values to override default settings
  // Don't change unless you know what you're doing
  //#define ETH_SPI_HOST        SPI3_HOST
  //#define SPI_CLOCK_MHZ       25
  
  // Must connect INT to GPIOxx or not working
  #define INT_GPIO            4    //9
  
  #define MISO_GPIO           37
  #define MOSI_GPIO           35
  #define SCK_GPIO            36
  #define CS_GPIO             34
```

Beware that `INT` is important. Without it, the library won't work
